File: scripts/predict_xgboost.py
#!/usr/bin/env python3
"""
XGBoost Prediction Script
Called by Node.js to make predictions
"""

import json
import sys
import xgboost as xgb
import numpy as np
from pathlib import Path

def load_model_and_metadata(model_dir):
    """Load XGBoost model and metadata"""
    model_dir = Path(model_dir)
    model_path = model_dir / 'model.json'
    metadata_path = model_dir / 'model-metadata.json'
    
    if not model_path.exists():
        raise FileNotFoundError(f"Model file not found: {model_path}")
    
    # Load model
    model = xgb.Booster()
    model.load_model(str(model_path))
    
    # Load metadata
    metadata = {}
    if metadata_path.exists():
        with open(metadata_path, 'r') as f:
            metadata = json.load(f)
    
    return model, metadata

def normalize_features(features, metadata):
    """Normalize features using stored normalization params"""
    # For now, we'll use the feature names from metadata
    # In production, you'd want to store normalization params
    feature_names = metadata.get('feature_names', [])
    
    # Extract features in the correct order
    feature_vector = [features.get(name, 0) for name in feature_names]
    
    return np.array([feature_vector])

def predict(features_json, model_dir):
    """Make prediction"""
    try:
        # Parse features
        features = json.loads(features_json)
        
        # Load model
        model, metadata = load_model_and_metadata(model_dir)
        
        # Prepare features
        X = normalize_features(features, metadata)
        
        # Make prediction
        dmatrix = xgb.DMatrix(X)
        prediction = model.predict(dmatrix)[0]
        
        # Ensure prediction is in [0, 1] range
        prediction = max(0.0, min(1.0, prediction))
        
        return prediction
    except Exception as e:
        raise Exception(f"Prediction error: {str(e)}")

if __name__ == '__main__':
    if len(sys.argv) < 3:
        print(json.dumps({'error': 'Usage: python3 predict_xgboost.py <model-dir> <features-json>'}))
        sys.exit(1)
    
    model_dir = sys.argv[1]
    features_json = sys.argv[2]
    
    try:
        quality = predict(features_json, model_dir)
        print(json.dumps({'predictedQuality': quality}))
    except Exception as e:
        print(json.dumps({'error': str(e)}))
        sys.exit(1)

File: scripts/predict-xgboost.js
#!/usr/bin/env node

/**
 * XGBoost Prediction Script
 * 
 * Called by mlModelIntegration to make predictions using XGBoost model
 * Uses Python XGBoost for predictions
 */

const { spawn } = require('child_process');
const path = require('path');
const fs = require('fs');

/**
 * Predict quality using XGBoost model
 */
function predictQuality(features, modelPath) {
  return new Promise((resolve, reject) => {
    const scriptPath = path.join(__dirname, 'predict_xgboost.py');
    const featuresJson = JSON.stringify(features);
    
    const python = spawn('python3', [scriptPath, modelPath, featuresJson]);
    
    let stdout = '';
    let stderr = '';
    
    python.stdout.on('data', (data) => {
      stdout += data.toString();
    });
    
    python.stderr.on('data', (data) => {
      stderr += data.toString();
    });
    
    python.on('close', (code) => {
      if (code !== 0) {
        reject(new Error(`Python script failed: ${stderr}`));
        return;
      }
      
      try {
        const result = JSON.parse(stdout.trim());
        resolve(result.predictedQuality);
      } catch (error) {
        reject(new Error(`Failed to parse prediction result: ${error.message}`));
      }
    });
  });
}

if (require.main === module) {
  const features = JSON.parse(process.argv[2] || '{}');
  const modelPath = process.argv[3];
  
  if (!modelPath) {
    console.error('Usage: node predict-xgboost.js <features-json> <model-path>');
    process.exit(1);
  }
  
  predictQuality(features, modelPath)
    .then(quality => {
      console.log(JSON.stringify({ predictedQuality: quality }));
    })
    .catch(error => {
      console.error('Error:', error.message);
      process.exit(1);
    });
}

module.exports = { predictQuality };

